src/sentry/relay/config/metric_extraction.py

```diff
@@ -0,0 +1,143 @@
+from dataclasses import dataclass
+from typing import Any, List, Literal, Sequence, TypedDict, Union
+
+from sentry.api.endpoints.project_transaction_threshold import DEFAULT_THRESHOLD
+from sentry.models import (
+    Project,
+    ProjectTransactionThreshold,
+    ProjectTransactionThresholdOverride,
+    TransactionMetric,
+)
+
+
+class RuleConditionInner(TypedDict):
+    op: Literal["eq", "gt"]
+    name: str
+    value: Any
+
+
+# mypy does not support recursive types. type definition is a very small subset
+# of the values relay actually accepts
+class RuleCondition(TypedDict):
+    op: Literal["and"]
+    inner: Sequence[RuleConditionInner]
+
+
+class MetricConditionalTaggingRule(TypedDict):
+    condition: RuleCondition
+    targetMetrics: Sequence[str]
+    targetTag: str
+    tagValue: str
+
+
+_TRANSACTION_METRICS_TO_RULE_FIELD = {
+    TransactionMetric.LCP.value: "transaction.measurements.lcp",
+    TransactionMetric.DURATION.value: "transaction.duration",
+}
+
+_SATISFACTION_TARGET_METRICS = (
+    "s:transactions/user@none",
+    "d:transactions/duration@millisecond",
+)
+
+_SATISFACTION_TARGET_TAG = "satisfaction"
+
+
+@dataclass
+class _DefaultThreshold:
+    metric: TransactionMetric
+    threshold: int
+
+
+_DEFAULT_THRESHOLD = _DefaultThreshold(
+    metric=TransactionMetric[DEFAULT_THRESHOLD["metric"].upper()].value,
+    threshold=int(DEFAULT_THRESHOLD["threshold"]),
+)
+
+
+def get_metric_conditional_tagging_rules(
+    project: Project,
+) -> Sequence[MetricConditionalTaggingRule]:
+    rules: List[MetricConditionalTaggingRule] = []
+    
+    # transaction-specific overrides must precede the project-wide threshold in the list of rules.
+    for threshold in project.projecttransactionthresholdoverride_set.all():
+        rules.extend(
+            _threshold_to_rules(
+                threshold,
+                [{"op": "eq", "name": "event.transaction", "value": threshold.transaction}],
+            )
+        )
+
+    # Rules are processed top-down. The following is a fallback for when
+    # there's no transaction-name-specific rule:
+
+    try:
+        threshold = ProjectTransactionThreshold.objects.get(project=project)
+        rules.extend(_threshold_to_rules(threshold, []))
+    except ProjectTransactionThreshold.DoesNotExist:
+        rules.extend(_threshold_to_rules(_DEFAULT_THRESHOLD, []))
+        pass
+
+    return rules
+
+
+def _threshold_to_rules(
+    threshold: Union[
+        ProjectTransactionThreshold, ProjectTransactionThresholdOverride, _DefaultThreshold
+    ],
+    extra_conditions: Sequence[RuleConditionInner],
+) -> Sequence[MetricConditionalTaggingRule]:
+    frustrated: MetricConditionalTaggingRule = {
+        "condition": {
+            "op": "and",
+            "inner": [
+                {
+                    "op": "gt",
+                    "name": _TRANSACTION_METRICS_TO_RULE_FIELD[threshold.metric],
```

Comment:
fwiw this was ported from this piece of relay code: https://github.com/getsentry/relay/blob/043479fe570949003964c672ed3dda7253a98292/relay-server/src/metrics_extraction/transactions.rs#L118-L132=